src/dune_rules/dune_file.ml

```diff
@@ -403,6 +403,7 @@ module Mode_conf = struct
       | Byte
       | Native
       | Best
+      | Melange
```

Comment:
I think this is used only for binaries which we aren't supporting yet.

Reply:
Agreed. I tried to remove it and it helps a lot removing edge cases where I was raising exns. However, parsing the `modes` stanza fails with 

Seems `Dune_file.parse_file_includes` is called while parsing the dune file? Here's the relevant stack trace (obtained by replacing user_error exn with regular one in `Decode.enum`):

```
Called from Dune_sexp__Decoder.parse in file "src/dune_sexp/decoder.ml", line
  249, characters 13-29
Called from Stdune__List.concat_map.aux in file "otherlibs/stdune/list.ml",
  line 41, characters 15-18
Called from Dune_rules__Dune_file.Stanzas.parse_file_includes in file
  "src/dune_rules/dune_file.ml", line 2177, characters 4-50
Called from Dune_rules__Dune_file.Stanzas.parse in file
  "src/dune_rules/dune_file.ml", line 2190, characters 6-55
Called from Dune_rules__Dune_file.parse in file
  "src/dune_rules/dune_file.ml", line 2239, characters 17-50
Called from Dune_rules__Dune_load.Dune_files.interpret.impl in file
  "src/dune_rules/dune_load.ml", line 202, characters 23-65
Called from Fiber__Scheduler.exec in file "src/fiber/scheduler.ml", line 73,
  characters 8-11
-> required by ("Dune_files.interpret", <opaque>)
```

Reply:
Never mind, it's the link mode that is used for binaries. This is the right change.

Reply:
> Here's the relevant stack trace (obtained by replacing user_error exn with regular one in Decode.enum):

You can run dune with `--debug-backtraces` to see the full trace.

Reply:
> it's the link mode that is used for binaries. This is the right change

The "mode conf" used for libraries and other cases was shared type for all cases. I was able to clean up link mode + all other cases by adding a new module `Mode_conf.Lib` that includes the melange mode, and is only used for libraries.

I am not 100% sure about it as it adds more boilerplate, but code looks cleaner now, and cases that should not be handled at the moment for melange (exes, installation, linking) are gone, please check https://github.com/jchavarri/dune/pull/4/commits/e8c901c43eb07c4931dbc42a587d18f64388b182

Reply:
Looks fine to me. I'd take straightforward boilerplate over imprecise types.